src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,413 @@
+            this.WritePendingMessageData(true);
+
+            this.SetState(BatchWriterState.BatchCompleted);
+
+            // write the end boundary for the batch
+            ODataMultipartMixedBatchWriterUtils.WriteEndBoundary(this.RawOutputContext.TextWriter, this.batchBoundary, !this.batchStartBoundaryWritten);
+
+            // For compatibility with WCF DS we write a newline after the end batch boundary.
+            // Technically it's not needed, but it doesn't violate anything either.
+            this.RawOutputContext.TextWriter.WriteLine();
+        }
+
+        /// <summary>
+        /// Ends an active changeset - implementation of the actual functionality.
+        /// </summary>
+        protected override void WriteEndChangesetImplementation()
+        {
+            // write pending message data (headers, response line) for a previously unclosed message/request
+            this.WritePendingMessageData(true);
+
+            string currentChangeSetBoundary = this.changeSetBoundary;
+
+            // change the state first so we validate the change set boundary before attempting to write it.
+            this.SetState(BatchWriterState.ChangesetCompleted);
+
+            Debug.Assert(this.changeSetBoundary != null, "this.changeSetBoundary != null");
+            this.changeSetBoundary = null;
+
+            // In the case of an empty changeset the start changeset boundary has not been written yet
+            // we will leave it like that, since we want the empty changeset to be represented only as
+            // the end changeset boundary.
+            // Due to WCF DS V2 compatibility we must not write the start boundary in this case
+            // otherwise WCF DS V2 won't be able to read it (it fails on the start-end boundary empty changeset).
+
+            // write the end boundary for the change set
+            ODataMultipartMixedBatchWriterUtils.WriteEndBoundary(this.RawOutputContext.TextWriter, currentChangeSetBoundary, !this.changesetStartBoundaryWritten);
+        }
+
+        /// <summary>
+        /// Creates an <see cref="ODataBatchOperationResponseMessage"/> for writing an operation of a batch response - implementation of the actual functionality.
+        /// </summary>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        /// <returns>The message that can be used to write the response operation.</returns>
+        protected override ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId)
+        {
+            this.WritePendingMessageData(true);
+
+            // In responses we don't need to use our batch URL resolver, since there are no cross referencing URLs
+            // so use the URL resolver from the batch message instead.
+            this.CurrentOperationResponseMessage = BuildOperationResponseMessage(
+                this.RawOutputContext.OutputStream, /*contentId*/ null, /*groupId*/ null);

```

Comment:
>/*groupId*/ null [](start = 72, length = 16)

for multipart/mixed, we should pass in the id used in generating the batchboundary.

Reply:
OK, also pass in the changeset id extracted from *changesetboundary* for groupId; contentId is directly using the value passed in. Please let me know any different thoughts.

---
In reply to: [156778151](https://github.com/OData/odata.net/pull/992#discussion_r156778151) [](ancestors = 156778151)